# Dataset

## Extract text from PDF

In [1]:
import os
import re
import json
import pandas as pd
import spacy
from PyPDF2 import PdfReader
import pdfplumber
from pathlib import Path
import logging
from typing import Dict, List, Any, Optional, Tuple, Set

In [2]:
nlp = spacy.load("en_core_web_lg")

## Get Data

- viewreport - good
- sample_report
- report 1
- report 2

In [146]:
pdf_path = "reports/sample report.pdf"

### Read PDF

In [147]:
with open(pdf_path,"rb") as f:
    reader = PdfReader(f)
    text2 = ""
    for page in reader.pages:
        text2 += page.extract_text() + "\n"
print(text2[:400])


District 8, 15 - LAMOILLE County
Owner: 1 - State Highway Agency
Town: 41 - CAMBRIDGE
Maintenance Responsibility: 1 - State Highway Agency
Team Lead:
 Stephen Piro,  
Inspection Date:
 09/27/2023
Structure 
#00020 /  
(Routine)
Route 
VT15 /  
VT 00015 ML over LAMOILLE RIVER

Copyright © 2024 Microsoft and its suppliers. All rights reserved.
Location: 0.9 MI E JCT. VT.104 N
44.64779, -72.87252
Tea


### Try fitz

In [148]:
import fitz  # PyMuPDF


document = fitz.open(pdf_path)
    
    # Variable to store the text
text3 = ""
    
    # Iterate through each page
for page_num in range(len(document)):
     # Get the page
    page = document.load_page(page_num)
        
        # Extract text from the page
    page_text = page.get_text()
        
        # Add the page text to our text variable
    text3 += page_text
    
    # Close the document
document.close()
    
text3

'District 8, 15 - LAMOILLE County\nOwner: 1 - State Highway Agency\nTown: 41 - CAMBRIDGE\nMaintenance Responsibility: 1 - State Highway Agency\nTeam Lead: Stephen Piro,  Inspection Date: 09/27/2023\nStructure #00020 /  (Routine)\nRoute VT15 /  \nVT 00015 ML over LAMOILLE RIVER\nCopyright © 2024 Microsoft and its suppliers. All rights reserved.\nLocation: 0.9 MI E JCT. VT.104 N\n44.64779, -72.87252\nTeam Lead: Stephen Piro,  Inspection Date: 09/27/2023\nStructure #00020 /  (Routine)\nRoute VT15 /  \nVT 00015 ML over LAMOILLE RIVER\nIDENTIFICATION\n(1) State Names\n50 - Vermont\n(8) Structure Number\n200030002008022\n(5) Inventory Route\n1\n(2) Highway Agency District\n8 - District 8\n(3) County Code\n15 - LAMOILLE\n(4) Place Code\n11425\n(6) Features Intersected\nLAMOILLE RIVER\n(7) Facility Carried\nVT 00015 ML\n(9) Location\n0.9 MI E JCT. VT.104 N\n(11) Mile Point\n24.576 mi\n(12) Base Highway Network\nYes\n(13) LRS Inventory Rte & Subrte\n0013100015\n(16) Latitude\n44.6477944444444\n

In [149]:
doc3 = nlp(text3)
doc3

District 8, 15 - LAMOILLE County
Owner: 1 - State Highway Agency
Town: 41 - CAMBRIDGE
Maintenance Responsibility: 1 - State Highway Agency
Team Lead: Stephen Piro,  Inspection Date: 09/27/2023
Structure #00020 /  (Routine)
Route VT15 /  
VT 00015 ML over LAMOILLE RIVER
Copyright © 2024 Microsoft and its suppliers. All rights reserved.
Location: 0.9 MI E JCT. VT.104 N
44.64779, -72.87252
Team Lead: Stephen Piro,  Inspection Date: 09/27/2023
Structure #00020 /  (Routine)
Route VT15 /  
VT 00015 ML over LAMOILLE RIVER
IDENTIFICATION
(1) State Names
50 - Vermont
(8) Structure Number
200030002008022
(5) Inventory Route
1
(2) Highway Agency District
8 - District 8
(3) County Code
15 - LAMOILLE
(4) Place Code
11425
(6) Features Intersected
LAMOILLE RIVER
(7) Facility Carried
VT 00015 ML
(9) Location
0.9 MI E JCT. VT.104 N
(11) Mile Point
24.576 mi
(12) Base Highway Network
Yes
(13) LRS Inventory Rte & Subrte
0013100015
(16) Latitude
44.6477944444444
(17) Longitude
-72.8725194444444
(98) Borde

### Try another

In [150]:
from pdfminer.high_level import extract_text

text4 = extract_text(pdf_path)
doc4 = nlp(text4)
doc4

CropBox missing from /Page, defaulting to MediaBox


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

Route VT15 /  

Structure #00020 /  (Routine)

VT 00015 ML over LAMOILLE RIVER

Team Lead: Stephen Piro,  Inspection Date: 09/27/2023

Town: 41 - CAMBRIDGE

District 8, 15 - LAMOILLE County

Owner: 1 - State Highway Agency

Maintenance Responsibility: 1 - State Highway Agency


Route VT15 /  

Structure #00020 /  (Routine)

VT 00015 ML over LAMOILLE RIVER

Team Lead: Stephen Piro,  Inspection Date: 09/27/2023

Location: 0.9 MI E JCT. VT.104 N

Copyright © 2024 Microsoft and its suppliers. All rights reserved.

44.64779, -72.87252


Route VT15 /  

Structure #00020 /  (Routine)

VT 00015 ML over LAMOILLE RIVER

Team Lead: Stephen Piro,  Inspection Date: 09/27/2023

IDENTIFICATION

CLASSIFICATION

(1) State Names
(8) Structure Number
(5) Inventory Route
(2) Highway Agency District
(3) County Code
(4) Place Code
(6) Features Intersected
(7) Facility Carried
(9) Location
(11) Mile Point
(12) Base Highway Network
(13) LRS Inventory Rte & Subrte
(16) Latitude
(17) Longitude
(98) Border Bridg

### Extract text better

In [151]:
text_with_tables = ""
tables = []
with pdfplumber.open(pdf_path) as pdf:
    for page in pdf.pages:
        # Extract tables first
        table_per_page = page.extract_tables()
        if(len(table_per_page) > 0):
            tables.append(table_per_page)
        # # Format tables as text
        for table in tables:
            for row in table:
                for a_row in row:
                    # print(a_row)
        #         # Filter None values and join cells with tabs
            #     print(row)
                    row_text = "\t".join([cell[0] if cell else "" for cell in a_row])
                    text_with_tables += row_text + "\n"
                text_with_tables += "\n"  # Extra line between tables
            
        #     # Extract the remaining text
        text = page.extract_text()
        if text:
            text_with_tables += text + "\n\n"
text_with_tables

CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

'Route VT15 /\nStructure #00020 / (Routine)\nVT 00015 ML over LAMOILLE RIVER\nTeam Lead: Stephen Piro, Inspection Date: 09/27/2023\nTown: 41 - CAMBRIDGE\nDistrict 8, 15 - LAMOILLE County\nOwner: 1 - State Highway Agency\nMaintenance Responsibility: 1 - State Highway Agency\n\nRoute VT15 /\nStructure #00020 / (Routine)\nVT 00015 ML over LAMOILLE RIVER\nTeam Lead: Stephen Piro, Inspection Date: 09/27/2023\nLocation: 0.9 MI E JCT. VT.104 N\nCopyright © 2024 Microsoft and its suppliers. All rights reserved.\n44.64779, -72.87252\n\nI\n(\n(\n(\nA\nB\nC\n*\n\nRoute VT15 /\nStructure #00020 / (Routine)\nVT 00015 ML over LAMOILLE RIVER\nTeam Lead: Stephen Piro, Inspection Date: 09/27/2023\nIDENTIFICATION CLASSIFICATION\n(1) State Names 50 - Vermont (112) NBIS Bridge Length Y\n(8) Structure Number 200030002008022 (104) Highway System 0\n(5) Inventory Route 1 (26) Functional Class 6 - Rural Minor Arterial\n(2) Highway Agency District 8 - District 8 (100) Defense Highway 1 - The inventory route is

In [152]:
print(tables)
len(tables)

[[[['INSPECTIONS *'], ['(90) Inspection Date 09/27/2023'], ['(91) Frequency 24'], ['(92) Critical Feature Inspection Done Freq. (Mon) Date'], ['A: Fracture Critical Detail No'], ['B: Underwater Inspection No'], ['C: Other Special Inspection'], ['* The inspection date and frequency information in this box contains\nthe current NBI date and frequency information. Please refer to the\nreport header for the date this inspection was conducted.']]], [[[None, None, None, None, None, '', '', ''], ['ELEMENTS', 'DESCRIPTION', 'UNITS', 'TOTAL', 'CS1', 'CS2', 'CS3', 'CS4'], ['12 Reinforced Concrete Deck SF 16740 16665 75 0 0\n1130 Cracking (RC and Other) SF 75 0 75 0 0\n510 Wearing Surfaces SF 13287 13287 0 0 0', None, None, None, None, None, None, None], ['301 Pourable Joint Seal LF 95 95 0 0 0', None, None, None, None, None, None, None], ['304 Open Expansion Joint LF 79 0 79 0 0\n2350 Debris Impaction LF 79 0 79 0 0', None, None, None, None, None, None, None], ['330 Metal Bridge Railing LF 812 7

7

### Clean Text

`re.sub()` this stands for subsititue 

In [153]:
text = re.sub(r'\s+', ' ', text_with_tables)  # Normalize whitespace
text = re.sub(r'(\d) - ', r'\1 - ', text_with_tables)
print(text[:400])

Route VT15 /
Structure #00020 / (Routine)
VT 00015 ML over LAMOILLE RIVER
Team Lead: Stephen Piro, Inspection Date: 09/27/2023
Town: 41 - CAMBRIDGE
District 8, 15 - LAMOILLE County
Owner: 1 - State Highway Agency
Maintenance Responsibility: 1 - State Highway Agency

Route VT15 /
Structure #00020 / (Routine)
VT 00015 ML over LAMOILLE RIVER
Team Lead: Stephen Piro, Inspection Date: 09/27/2023
Locati


In [154]:
doc = nlp(text)
# doc

In [155]:
text2 = re.sub(r'\s+', ' ', text2)  # Normalize whitespace
text2 = re.sub(r'(\d) - ', r'\1 - ', text2)
print(text2[:400])

District 8, 15 - LAMOILLE County Owner: 1 - State Highway Agency Town: 41 - CAMBRIDGE Maintenance Responsibility: 1 - State Highway Agency Team Lead: Stephen Piro, Inspection Date: 09/27/2023 Structure #00020 / (Routine) Route VT15 / VT 00015 ML over LAMOILLE RIVER Copyright © 2024 Microsoft and its suppliers. All rights reserved. Location: 0.9 MI E JCT. VT.104 N 44.64779, -72.87252 Team Lead: Ste


In [156]:
doc2 = nlp(text2)
# doc2

In [157]:
doc

Route VT15 /
Structure #00020 / (Routine)
VT 00015 ML over LAMOILLE RIVER
Team Lead: Stephen Piro, Inspection Date: 09/27/2023
Town: 41 - CAMBRIDGE
District 8, 15 - LAMOILLE County
Owner: 1 - State Highway Agency
Maintenance Responsibility: 1 - State Highway Agency

Route VT15 /
Structure #00020 / (Routine)
VT 00015 ML over LAMOILLE RIVER
Team Lead: Stephen Piro, Inspection Date: 09/27/2023
Location: 0.9 MI E JCT. VT.104 N
Copyright © 2024 Microsoft and its suppliers. All rights reserved.
44.64779, -72.87252

I
(
(
(
A
B
C
*

Route VT15 /
Structure #00020 / (Routine)
VT 00015 ML over LAMOILLE RIVER
Team Lead: Stephen Piro, Inspection Date: 09/27/2023
IDENTIFICATION CLASSIFICATION
(1) State Names 50 - Vermont (112) NBIS Bridge Length Y
(8) Structure Number 200030002008022 (104) Highway System 0
(5) Inventory Route 1 (26) Functional Class 6 - Rural Minor Arterial
(2) Highway Agency District 8 - District 8 (100) Defense Highway 1 - The inventory route is on
(3) County Code 15 - LAMOILLE (

In [158]:
pattern = f"Structure\s+#(\d+)"
matches = re.finditer(pattern, doc.text, re.IGNORECASE)
structure_numbers = []
for match in matches:
    if match.group(1) not in structure_numbers:
        structure_numbers.append(match.group(1))
bridge_id = structure_numbers[0] if len(structure_numbers) == 1 else "Unknown"
bridge_id

<>:1: SyntaxWarning: invalid escape sequence '\s'
<>:1: SyntaxWarning: invalid escape sequence '\s'
C:\Users\MK\AppData\Local\Temp\ipykernel_7200\608127002.py:1: SyntaxWarning: invalid escape sequence '\s'
  pattern = f"Structure\s+#(\d+)"


'00020'

## Find section boundaries

In [159]:
sentences = list(doc.sents) # this is how you get the sentences
sentences[0].start_char

0

In [161]:
# senteences = list(doc.sents)
lines = doc.text.split('\n')
section_boundaries = []

section_headers = [
        "deck", "approach", "superstructure", "substructure", 
        "channel", "general observation","channel profile",
    ]
for i, line in enumerate(lines):
    clean_line = line.strip().lower()
    if not clean_line:
        continue
    for header in section_headers:
        if clean_line == header:
            section_boundaries.append((i, header.upper()))
            break
section_boundaries.sort(key=lambda x: x[0])
print(section_boundaries)

[(124, 'DECK'), (188, 'APPROACH'), (227, 'SUPERSTRUCTURE'), (278, 'SUBSTRUCTURE'), (306, 'CHANNEL'), (352, 'GENERAL OBSERVATION'), (513, 'CHANNEL PROFILE'), (762, 'SUPERSTRUCTURE')]


In [162]:
section= {}
sections = []

for i, (line_idx, header) in enumerate(section_boundaries):
    end_idx = section_boundaries[i + 1][0] if  i < len(section_boundaries) - 1 else len(lines)
    
    section = {
        "index": (line_idx, end_idx),
        "content_array": lines[line_idx:end_idx],
        "content": "\n".join(lines[line_idx:end_idx]),
        "header": header
    }
    sections.append(section)

In [353]:
print(sections[3]["content"])

Substructure
ELEMENTS DESCRIPTION UNITS TOTAL CS1 CS2 CS3 CS4
205 Reinforced Concrete Column EA 7 7 0 0 0
210 Reinforced Concrete Pier Wall LF 175 11 139 25 0
1080 Delamination/Spall/Patched Area LF 65 0 65 0 0
1120 Efflorescence/Rust Staining LF 55 0 30 25 0
1190 Abrasion/Wear (PSC/RC) LF 44 0 44 0 0
234 Reinforced Concrete Pier Cap LF 171 126 0 45 0
1120 Efflorescence/Rust Staining LF 45 0 0 45 0
60 - Substructure (7 - GOOD CONDITION - some minor problems.)
Reinforced concrete abutment #1 is in fairly good condition having some slight undermining between the columns that
penetrates to the backside along the downstream side and less along the upstream side.
Reinforced concrete abutment #2 is in fairly good condition.
A71 - Abutment End Walls Condition (Good)
Reinforced concrete curtain walls are present over both abutments are in fairly good condition.
A77 - Retaining/Wingwall Condition (Good)
Concrete wingwalls are in fairly good condition.
A81 - Pier Seat/Cap Condition (Good)
Reinfo

## Get table and description boundary

we will use units here because these are present for everyt row 

In [354]:
units_patterns = [
        r'\bLF\b',    # Linear Feet
        r'\bEA\b',    # Each
        r'\bSF\b',    # Square Feet
        r'\bCF\b',    # Cubic Feet
        r'\bFT\b',    # Feet 
        r'\bIN\b',    # Inches
        r'\bM\b',     # Meters
        r'\bMM\b'     # Millimeters
    ]
units_regex = '|'.join(units_patterns)

In [355]:
units_regex

'\\bLF\\b|\\bEA\\b|\\bSF\\b|\\bCF\\b|\\bFT\\b|\\bIN\\b|\\bM\\b|\\bMM\\b'

A simple check to asses if content has a table or not

In [356]:
if "LF" in sections[3]["content"]:
    print("Yeah")
else:
    print("Nah")

Yeah


### Get last table line

In [357]:
doc_ = nlp(sections[3]["content"])
lines = doc_.text.split("\n")

last_table_line = -1
for idx, line in enumerate(lines):
    # print(re.search(units_regex, line))
    if re.search(units_regex, line):
        print(line)
        last_table_line = idx

205 Reinforced Concrete Column EA 7 7 0 0 0
210 Reinforced Concrete Pier Wall LF 175 11 139 25 0
1080 Delamination/Spall/Patched Area LF 65 0 65 0 0
1120 Efflorescence/Rust Staining LF 55 0 30 25 0
1190 Abrasion/Wear (PSC/RC) LF 44 0 44 0 0
234 Reinforced Concrete Pier Cap LF 171 126 0 45 0
1120 Efflorescence/Rust Staining LF 45 0 0 45 0


In [358]:
lines

['Substructure',
 'ELEMENTS DESCRIPTION UNITS TOTAL CS1 CS2 CS3 CS4',
 '205 Reinforced Concrete Column EA 7 7 0 0 0',
 '210 Reinforced Concrete Pier Wall LF 175 11 139 25 0',
 '1080 Delamination/Spall/Patched Area LF 65 0 65 0 0',
 '1120 Efflorescence/Rust Staining LF 55 0 30 25 0',
 '1190 Abrasion/Wear (PSC/RC) LF 44 0 44 0 0',
 '234 Reinforced Concrete Pier Cap LF 171 126 0 45 0',
 '1120 Efflorescence/Rust Staining LF 45 0 0 45 0',
 '60 - Substructure (7 - GOOD CONDITION - some minor problems.)',
 'Reinforced concrete abutment #1 is in fairly good condition having some slight undermining between the columns that',
 'penetrates to the backside along the downstream side and less along the upstream side.',
 'Reinforced concrete abutment #2 is in fairly good condition.',
 'A71 - Abutment End Walls Condition (Good)',
 'Reinforced concrete curtain walls are present over both abutments are in fairly good condition.',
 'A77 - Retaining/Wingwall Condition (Good)',
 'Concrete wingwalls are in 

In [359]:
lines[last_table_line]

'1120 Efflorescence/Rust Staining LF 45 0 0 45 0'

In [360]:

table = lines[:last_table_line+1]
desc = lines[last_table_line+1:]
table, desc

(['Substructure',
  'ELEMENTS DESCRIPTION UNITS TOTAL CS1 CS2 CS3 CS4',
  '205 Reinforced Concrete Column EA 7 7 0 0 0',
  '210 Reinforced Concrete Pier Wall LF 175 11 139 25 0',
  '1080 Delamination/Spall/Patched Area LF 65 0 65 0 0',
  '1120 Efflorescence/Rust Staining LF 55 0 30 25 0',
  '1190 Abrasion/Wear (PSC/RC) LF 44 0 44 0 0',
  '234 Reinforced Concrete Pier Cap LF 171 126 0 45 0',
  '1120 Efflorescence/Rust Staining LF 45 0 0 45 0'],
 ['60 - Substructure (7 - GOOD CONDITION - some minor problems.)',
  'Reinforced concrete abutment #1 is in fairly good condition having some slight undermining between the columns that',
  'penetrates to the backside along the downstream side and less along the upstream side.',
  'Reinforced concrete abutment #2 is in fairly good condition.',
  'A71 - Abutment End Walls Condition (Good)',
  'Reinforced concrete curtain walls are present over both abutments are in fairly good condition.',
  'A77 - Retaining/Wingwall Condition (Good)',
  'Concrete

In [361]:
table_text = '\n'.join(table)
print(table_text)

Substructure
ELEMENTS DESCRIPTION UNITS TOTAL CS1 CS2 CS3 CS4
205 Reinforced Concrete Column EA 7 7 0 0 0
210 Reinforced Concrete Pier Wall LF 175 11 139 25 0
1080 Delamination/Spall/Patched Area LF 65 0 65 0 0
1120 Efflorescence/Rust Staining LF 55 0 30 25 0
1190 Abrasion/Wear (PSC/RC) LF 44 0 44 0 0
234 Reinforced Concrete Pier Cap LF 171 126 0 45 0
1120 Efflorescence/Rust Staining LF 45 0 0 45 0


In [362]:
desc_text = '\n'.join(desc)
print(desc_text)

60 - Substructure (7 - GOOD CONDITION - some minor problems.)
Reinforced concrete abutment #1 is in fairly good condition having some slight undermining between the columns that
penetrates to the backside along the downstream side and less along the upstream side.
Reinforced concrete abutment #2 is in fairly good condition.
A71 - Abutment End Walls Condition (Good)
Reinforced concrete curtain walls are present over both abutments are in fairly good condition.
A77 - Retaining/Wingwall Condition (Good)
Concrete wingwalls are in fairly good condition.
A81 - Pier Seat/Cap Condition (Good)
Reinforced concrete pier caps are in fairly good condition having scattered various size rust stains, efflorescence leakage
and minor cracking.
A83 - Pier Shaft Condition (Good)
Reinforced concrete pier shafts are in fairly good condition having scattered previously patched areas along the pier walls
with various size rust stains scattered along the faces. Lower portions of each pier has minor abrasion al

## Get bridge components from tables

### Create component regex

In [363]:
processed_components =[]
bridge_components = [
    "Deck",
    "Open Girder/Beam",
    "Girder/Beam",
    "Pourable Joint Seal",
    "Open Expansion Joint",
    "Truss",
    "Column",
    "Arch",
    "Pier",
    "Pier Wall",
    "Abutment",
    "Bearing",
    "Pile",
    "Railing",
    "Pile Cap",
    "Bent",
    "Parapet/Guard Rail",
    "Expansion Joint",
    "Foundation",
    "Backwall",
    "Diaphragm",
    "Fascia",
    "Approach",
    "Slab",
    "Approach Slab",
    "Wingwall",
    "Spandrel",
    "Cantilever",
    "Movable Bearing",
    "Fixed Bearing",
    "Counterweight",
    "Cutwater",
    "Flood Arches",
    "Suspension Cable",
    "Tower",
    "Cable-Stay",
    "Wing/Retaining Wall",
    "Pier Cap",
    "Bearing Pedestal",
    "Stringer",
    "Footing",
    "Shear Key",
    "Drainage System",
    "Seismic Device",
    "Crash Barrier",
    "Culvert",
    "Culvert Pipe",
    "Headwall",
    "Apron",
    "Invert",
    "Soffit",
    "Skewback",
    "Keystone",
    "Camber",
    "Tie Rod",
    "Dampener"
]
bridge_components = list(set(bridge_components))
materials = ["Reinforced Concrete", "Prestressed Concrete", "RC", "PC", "Steel", "Concrete", "Timber", "Wood", "Metal", "Metal Bridge"]


In [364]:
for component in bridge_components:
        # Handle slash-separated components
        if "/" in component:
            parts = [re.escape(part.strip()) for part in component.split("/")]
            # Create pattern that allows for typos or partial matches
            pattern = f"(?:{'|'.join(parts)}|{re.escape(component)})"
            
            # Special case for "Wing/Retaining Wall" - also match if "Wall" has a typo
            if "Wall" in component:
                wall_variations = ["Wall", "Wal", "Walls"]
                wall_pattern = f"(?:{'|'.join(wall_variations)})"
                # Replace "Wall" in the pattern with the variations
                pattern = pattern.replace(re.escape("Wall"), wall_pattern)
        else:
            pattern = re.escape(component)
        
        # Make trailing 's' optional for plurals (e.g. match "Piers" with "Pier")
        if pattern.endswith("s\\"):
            pattern = pattern[:-2] + "s?"
        else:
            pattern = pattern + "s?"
            
        # Remove number if present (e.g. "Abutment 1" -> "Abutment( \d+)?")
        if " " in pattern and any(c.isdigit() for c in pattern):
            pattern = re.sub(r'\\\ \d+', '', pattern) + r'(\ \d+)?'
            
        processed_components.append(pattern)

In [365]:
material_pattern = f"(?:{'|'.join(materials)})\\s+"
component_pattern = f"(?:{'|'.join(processed_components)})"
    
# Final pattern including optional material prefix and component
final_pattern = fr"(\d+\s+(?:{material_pattern})?{component_pattern})"

In [366]:
component_sections = []
doc_t = nlp(table_text)
    
matches = list(re.finditer(final_pattern, doc_t.text))
for i, match in enumerate(matches):
    start_pos = match.start()
        
    # Determine end position (either next component or end of text)
    if i < len(matches) - 1:
        end_pos = matches[i+1].start()
    else:
        end_pos = len(table_text)
        
    # Extract the full section text
    section_text = table_text[start_pos:end_pos].strip()
        
    # Extract just the component name (may include material)
    component_text = match.group(1)
        
    component_sections.append({
            "full_section": section_text,
            "component": component_text,
            "start": start_pos,
            "end": end_pos
        })

In [367]:
print(component_sections[0]["full_section"])

205 Reinforced Concrete Column EA 7 7 0 0 0


In [368]:
print(component_sections[1]["full_section"])

210 Reinforced Concrete Pier Wall LF 175 11 139 25 0
1080 Delamination/Spall/Patched Area LF 65 0 65 0 0
1120 Efflorescence/Rust Staining LF 55 0 30 25 0
1190 Abrasion/Wear (PSC/RC) LF 44 0 44 0 0


In [369]:
print(component_sections[2]["full_section"])

234 Reinforced Concrete Pier Cap LF 171 126 0 45 0
1120 Efflorescence/Rust Staining LF 45 0 0 45 0


## Get Component and defect Info

### Component info

In [370]:
# Regex patterns for extracting data
id_pattern = r"^(\d+)\s+"
units_pattern = r"\b(LF|SF|EA|CF|FT|IN)\b"
values_pattern = r"\b(\d+(?:\s+\d+)*)\s*$"

In [371]:
def extract_component_values(component_text, full_section, component_data):
    """
    Extract component values directly after the component name and unit.
    
    Args:
        component_text (str): The component text (e.g. "215 Reinforced Concrete Abutment")
        full_section (str): The full text of the component section
        component_data (dict): The component data dictionary to update
    """
    # Escape special characters in the component text for regex
    escaped_component = re.escape(component_text)
    
    # Pattern to find the component's unit and values
    # Look for exactly 5 numbers after the unit (TOTAL + CS1-CS4)
    pattern = fr"{escaped_component}\s+(LF|SF|EA|CF|FT|IN)\s+(\d+\s+\d+\s+\d+\s+\d+\s+\d+)"
    match = re.search(pattern, full_section)
    
    if match:
        component_data["units"] = match.group(1)
        values_str = match.group(2)
        component_data["values"] = [int(val) for val in values_str.split()]
    else:
        # Fallback pattern for cases with fewer or more values
        pattern = fr"{escaped_component}\s+(LF|SF|EA|CF|FT|IN)\s+(\d+(?:\s+\d+)*?)(?:\s+\d{{3,4}}\s|$)"
        match = re.search(pattern, full_section)
        
        if match:
            component_data["units"] = match.group(1)
            values_str = match.group(2)
            component_data["values"] = [int(val) for val in values_str.split()]
        else:
            # Last resort fallback
            unit_pattern = r"\b(LF|SF|EA|CF|FT|IN)\s+(\d+(?:\s+\d+){0,4})"
            pos = full_section.find(component_text) + len(component_text)
            remaining_text = full_section[pos:pos+30]  # Look at limited text to avoid capturing too much
            unit_match = re.search(unit_pattern, remaining_text)
            
            if unit_match:
                component_data["units"] = unit_match.group(1)
                values_str = unit_match.group(2)

                component_data["values"] = [int(val) for val in values_str.split()]

In [372]:
def extract_defects(section_text, component_data):
    """
    Extract defect information from a component section.
    
    Args:
        section_text (str): The full component section text
        component_data (dict): The component data dictionary to update
    """
    # Pattern to match defect entries: ID + Description + Units + Values
    # Use a more specific pattern to match exactly the format we expect
    defect_pattern = r"(\d{3,4})\s+([^L][^F][^S][^E][^A][^\d]+?)\s+(LF|SF|EA|CF|FT|IN)\s+(\d+(?:\s+\d+){0,4})"
    
    # Find all defects in the section text
    defect_matches = re.finditer(defect_pattern, section_text)
    
    for match in defect_matches:
        defect_id = match.group(1)
        defect_desc = match.group(2).strip()
        defect_unit = match.group(3)
        defect_values_str = match.group(4)
        defect_values = [int(val) for val in defect_values_str.split()]
        
        # Skip if this is the component ID itself (first entry)
        if defect_id == component_data.get("id"):
            continue
            
        # Create defect entry
        defect = {
            "id": defect_id,
            "description": defect_desc,
            "unit": defect_unit,
            "total": defect_values[0] if defect_values else None,
            "cs_values": defect_values[1:] if len(defect_values) > 1 else []
        }
        
        component_data["defects"].append(defect)

In [373]:
structured_components = []
for section in component_sections:
    component_data = {
            "text": section["component"],
            "full_section": section["full_section"],
            "defects": []  # Will hold defect information
        }
        
    # Extract ID
    id_match = re.search(id_pattern, section["component"])
    if id_match:
        print(id_match.group(1))
        component_data["id"] = id_match.group(1)
    units_match = re.search(units_pattern, section["full_section"])
    if units_match:
        print(units_match.group(1))
        component_data["units"] = units_match.group(1)
    values_match = re.search(values_pattern, section["full_section"])
    extract_component_values(section["component"], section["full_section"], component_data)
    extract_defects(section["full_section"], component_data)
    structured_components.append(component_data)

205
EA
210
LF
234
LF


Expetected Output
`[{'text': '215 Reinforced Concrete Abutment',
  'full_section': '215 Reinforced Concrete Abutment LF 51 42 6 3 0\n1080 Delamination/Spall/Patched Area LF 3 0 0 3 0\n1090 Exposed Rebar LF 1 0 1 0 0\n1130 Cracking (RC and Other) LF 5 0 5 0 0',
  'defects': [{'id': '1080',
    'description': 'Delamination/Spall/Patched Area',
    'unit': 'LF',
    'total': 3,
    'cs_values': [0, 0, 3, 0]},
   {'id': '1090',
    'description': 'Exposed Rebar',
    'unit': 'LF',
    'total': 1,
    'cs_values': [0, 1, 0, 0]},
   {'id': '1130',
    'description': 'Cracking (RC and Other)',
    'unit': 'LF',
    'total': 5,
    'cs_values': [0, 5, 0, 0]}],
  'id': '215',
  'units': 'LF',
  'values': [51, 42, 6, 3, 0]},
 {'text': '800 Reinforced Concrete Wing',
  'full_section': '800 Reinforced Concrete Wing/Retaining Wall EA 4 0 4 0 0\n1120 Efflorescence/Rust Staining EA 2 0 2 0 0\n1130 Cracking (RC and Other) EA 2 0 2 0 0',
  'defects': [{'id': '1120',
    'description': 'Efflorescence/Rust Staining',
    'unit': 'EA',
    'total': 2,
    'cs_values': [0, 2, 0, 0]},
   {'id': '1130',
    'description': 'Cracking (RC and Other)',
    'unit': 'EA',
    'total': 2,
    'cs_values': [0, 2, 0, 0]}],
  'id': '800',
  'units': 'EA',
  'values': [4, 0, 4, 0, 0]}]`

In [374]:
structured_components

[{'text': '205 Reinforced Concrete Column',
  'full_section': '205 Reinforced Concrete Column EA 7 7 0 0 0',
  'defects': [],
  'id': '205',
  'units': 'EA',
  'values': [7, 7, 0, 0, 0]},
 {'text': '210 Reinforced Concrete Pier Wall',
  'full_section': '210 Reinforced Concrete Pier Wall LF 175 11 139 25 0\n1080 Delamination/Spall/Patched Area LF 65 0 65 0 0\n1120 Efflorescence/Rust Staining LF 55 0 30 25 0\n1190 Abrasion/Wear (PSC/RC) LF 44 0 44 0 0',
  'defects': [{'id': '1080',
    'description': 'Delamination/Spall/Patched Area',
    'unit': 'LF',
    'total': 65,
    'cs_values': [0, 65, 0, 0]},
   {'id': '1120',
    'description': 'Efflorescence/Rust Staining',
    'unit': 'LF',
    'total': 55,
    'cs_values': [0, 30, 25, 0]},
   {'id': '1190',
    'description': 'Abrasion/Wear (PSC/RC)',
    'unit': 'LF',
    'total': 44,
    'cs_values': [0, 44, 0, 0]}],
  'id': '210',
  'units': 'LF',
  'values': [175, 11, 139, 25, 0]},
 {'text': '234 Reinforced Concrete Pier',
  'full_sectio

## Get info from description text

### get contexts

In [303]:
desc_text.split("\n")

['59 - Superstructure (6 - SATISFACTORY CONDITION - structural elements show some minor deterioration.)',
 'Seven (7) painted steel rolled beams are in satisfactory condition having scattered minor to moderate previous pitting',
 'along the flanges and webs throughout. Superstructure was painted in 2008 and is holding up fairly well with a few areas',
 'where the top layer of paint is starting to peel.',
 'A55 - Lateral Bracing Condition (Good)',
 'Spans #1 and #5 have five (5) painted steel c-channels while remaining spans have six (6) painted steel c-channels',
 'present per bay that are welded to steel T-plates that are welded to the webs of the rolled beams are in fairly good',
 'condition.',
 'B.C.07 Bridge Bearings Condition Rating (GOOD - Some minor defects.)',
 'Fabric bearings are in good condition.',
 '',
 'I',
 '(',
 '(',
 '(',
 'A',
 'B',
 'C',
 '*',
 '',
 '\t\t\t\t\t\t\t',
 'E\tD\tU\tT\tC\tC\tC\tC',
 '1\t\t\t\t\t\t\t',
 '3\t\t\t\t\t\t\t',
 '3\t\t\t\t\t\t\t',
 '3\t\t\t\t\t\

**Important** - you added this ti make sure there is not much overflow double check this change

In [304]:
cleaned_idx = desc_text.split("\n").index('')
desc_text = "\n".join(desc_text.split("\n")[:cleaned_idx])

In [305]:
def parse_condition(condition_text):
    """
    Parse condition text to extract rating and assessment description.
    
    Args:
        condition_text (str): The condition text from the header
        
    Returns:
        tuple: (condition_rating, assessment_description)
    """
    if not condition_text:
        return None, None
    
    # Handle "N/A" or "Not Applicable" cases
    if condition_text.upper() in ["N/A", "NOT APPLICABLE"]:
        return "Not Applicable", None
    
    # Handle simple condition ratings without explanations
    if condition_text.upper() in ["SATISFACTORY", "GOOD", "FAIR", "POOR"]:
        return condition_text.upper(), None
    
    # Pattern for conditions with numeric codes: "6 - SATISFACTORY CONDITION - explanation"
    code_pattern = r'(?:(\d+)\s*-\s*)?([A-Z\s]+)(?:\s*-\s*(.+))?'
    match = re.match(code_pattern, condition_text)
    
    if match:
        code = match.group(1)  # Numeric code like "6"
        rating = match.group(2).strip()  # Rating like "SATISFACTORY CONDITION"
        assessment = match.group(3)  # Explanation text
        
        # Normalize rating text (remove "CONDITION" if present)
        rating = rating.replace("CONDITION", "").strip()
        
        return rating, assessment
    
    # Default case - return the whole text as the rating
    return condition_text, None

In [203]:
# header_pattern = r'([A-Z]?\d+\s+-\s+[^()]+)(?:\s*\(([^)]+)\))?'
    
#     # Find all headers in the text
# header_matches = list(re.finditer(header_pattern, desc_text))
    
#     # If no headers found, return the whole text as one section
# if not header_matches:
#     desc_sections =  [{"header": None, "condition": None, "content": text.strip()}]
    
#     # Create sections by finding the text between headers
# desc_sections = []
    
# for i, match in enumerate(header_matches):
#     header = match.group(1).strip()
#     condition_text = match.group(2).strip() if match.group(2) else None
        
#         # Parse the condition and assessment
#     condition_rating, assessment = parse_condition(condition_text)
        
#         # Find the start of the content (after the header and condition)
#     content_start = match.end()
        
#         # Find the end of the content (start of next header or end of text)
#     if i < len(header_matches) - 1:
#         content_end = header_matches[i + 1].start()
#     else:
#         content_end = len(text)
        
#         # Extract the content
#     content = desc_text[content_start:content_end].strip()
        
#     desc_sections.append({
#             "header": header,
#             "condition_text": condition_text,
#             "condition_rating": condition_rating,
#             "assessment": assessment,
#             "content": content
#         })
# desc_sections

**Important** --- Retry this code make sure it works with substructure

In [306]:
header_pattern = r'(?:([A-Z](?:\.[A-Z])?(?:\.\d+)?(?:\s+-)?|\d+\s+-)\s+([^()]+))(?:\s*\(([^)]+)\))?'

header_matches = list(re.finditer(header_pattern, desc_text))
    
    # If no headers found, return the whole text as one section
if not header_matches:
    desc_sections = [{"header": None, "condition_text": None, "condition_rating": None, 
                "assessment": None, "content": text.strip()}]
    
    # Create sections by finding the text between headers
desc_sections = []
    
for i, match in enumerate(header_matches):
        # Combine the prefix and title parts to form the complete header
    prefix = match.group(1).strip() if match.group(1) else ""
    title = match.group(2).strip() if match.group(2) else ""
    header = f"{prefix} {title}".strip()
        
    condition_text = match.group(3).strip() if match.group(3) else None
        
        # Parse the condition and assessment
    condition_rating, assessment = parse_condition(condition_text)
        
        # Find the start of the content (after the header and condition)
    content_start = match.end()
        
        # Find the end of the content (start of next header or end of text)
    if i < len(header_matches) - 1:
        content_end = header_matches[i + 1].start()
    else:
        content_end = len(text)
        
        # Extract the content
    content = desc_text[content_start:content_end].strip()
        
    desc_sections.append({
            "header": header,
            "condition_text": condition_text,
            "condition_rating": condition_rating,
            "assessment": assessment,
            "content": content if len(content) > 5 else ""
        })
    
desc_sections

[{'header': '59 - Superstructure',
  'condition_text': '6 - SATISFACTORY CONDITION - structural elements show some minor deterioration.',
  'condition_rating': 'SATISFACTORY',
  'assessment': 'structural elements show some minor deterioration.',
  'content': 'Seven (7) painted steel rolled beams are in satisfactory condition having scattered minor to moderate previous pitting\nalong the flanges and webs throughout. Superstructure was painted in 2008 and is holding up fairly well with a few areas\nwhere the top layer of paint is starting to peel.\nA'},
 {'header': '55 - Lateral Bracing Condition',
  'condition_text': 'Good',
  'condition_rating': 'GOOD',
  'assessment': None,
  'content': 'Spans #1 and #5 have five (5) painted steel c-channels while remaining spans have six (6) painted steel c-channels\npresent per bay that are welded to steel T-plates that are welded to the webs of the rolled beams are in fairly good\ncondition.'},
 {'header': 'B.C.07 Bridge Bearings Condition Rating',

## Matching components and desciption

### Extract component name

remeber here we would neeed to extend to specoal cases

In [324]:
def extract_component_name(text):
    """
    Extract the core component name from the text, removing ID and material prefixes.
    
    Args:
        text (str): The component text (e.g., "215 Reinforced Concrete Abutment")
        
    Returns:
        str: The core component name (e.g., "Abutment")
    """
    # Remove leading ID number
    text = re.sub(r'^\d+\s+', '', text)
    
    # Remove common material prefixes
    for material in materials:
        text = text.replace(material, '').strip()

    # we will need to extend special cases
    if text == "Open Girder" or text == "Girder":
        text = "Beam"
    if "Bearing" in text:
        text = "Bearing"
    
    # Handle special case for "Wing/Retaining Wall"
    if "Wing" in text and "Wall" not in text:
        text = "Wing/Retaining Wall"
    
    return text

In [325]:
len(structured_components)

3

In [327]:
print(structured_components[2]['text'])
extract_component_name(structured_components[2]['text'])

313 Fixed Bearing


'Bearing'

### Generate component variations

what about lowercase with s and no s?


Rememer here we would dstill need to special cases

In [328]:
def generate_component_variations(component_name):
    """
    Generate variations of a component name for flexible matching.
    
    Args:
        component_name (str): The core component name
        
    Returns:
        list: List of possible variations of the component name
    """
    variations = [component_name]
    
    # Add lowercase version
    variations.append(component_name.lower())
    
    # Handle singular/plural variations
    if component_name.endswith('s'):
        variations.append(component_name[:-1])  # Remove 's'
    else:
        variations.append(component_name + 's')  # Add 's'
    
    # Handle variations with numbers
    variations.append(component_name + r' \d+')  # Component followed by number
    
    # Handle compound components with slashes
    if '/' in component_name:
        parts = [part.strip() for part in component_name.split('/')]
        variations.extend(parts)
        
        # Special case for "Wing/Retaining Wall"
        if "Wing" in component_name and "Wall" in component_name:
            variations.append("Wingwall")
    return variations

In [330]:
print(extract_component_name(structured_components[0]['text']))
generate_component_variations(extract_component_name(structured_components[0]['text']))

Beam


['Beam', 'beam', 'Beams', 'Beam \\d+']

### Match comoponent name in headers

**Important** -- you changed code here make sure it works for older pdfs

In [331]:
def find_in_headers(component_variations, description_sections):
    """
    Try to find component variations in the section headers.
    
    Args:
        component_variations (list): List of component name variations
        description_sections (list): List of section dictionaries
        
    Returns:
        dict or None: The matching section or None if no match found
    """
    for section in description_sections:
        header = section["header"]
        
        if not header:  # to check if a string exists in a larger string just do this 
            continue
        # print(header)
        if "joint" in header.lower() and "trough" in header.lower():
            # print("here")
            header = "Open Expansion Joint"
            # print(header)    
        elif "joint" in header.lower() and "bridge" in header.lower():
            header = "Pourable Joint Seal"
        # Check if any variation appears in the header
        for variation in component_variations:
            # print(variation)
            if variation.lower() in header.lower():
                # print("here 2")
                return section
    
    return None

In [332]:
def check_comp_desc_matching(idx):
    print(structured_components[idx]['text'])
    return find_in_headers(generate_component_variations(extract_component_name(structured_components[idx]['text'])), desc_sections)

In [335]:
check_comp_desc_matching(0)

107 Steel Open Girder


### Find in content

Double check  what is going here there might be an issue

In [336]:
def find_in_content(component_variations, description_sections):
    """
    Try to find component variations in the section content.
    
    Args:
        component_variations (list): List of component name variations
        description_sections (list): List of section dictionaries
        
    Returns:
        dict or None: The matching section or None if no match found
    """
    for section in description_sections:
        content = section["content"]
        
        if not content:
            continue
            
        # Check if any variation appears in the content
        for variation in component_variations:
            # Create a pattern that matches the variation as a whole word
            pattern = r'\b' + variation + r'\b'
            if re.search(pattern, content, re.IGNORECASE):
                return section
                
            # For patterns with \d+, we need a special approach
            if r'\d+' in variation:
                base = variation.replace(r'\d+', '')
                pattern = r'\b' + base + r'\d+\b'
                if re.search(pattern, content, re.IGNORECASE):
                    return section
    
    return None

In [337]:
find_in_content(generate_component_variations(extract_component_name(structured_components[0]['text'])), desc_sections)

{'header': '59 - Superstructure',
 'condition_text': '6 - SATISFACTORY CONDITION - structural elements show some minor deterioration.',
 'condition_rating': 'SATISFACTORY',
 'assessment': 'structural elements show some minor deterioration.',
 'content': 'Seven (7) painted steel rolled beams are in satisfactory condition having scattered minor to moderate previous pitting\nalong the flanges and webs throughout. Superstructure was painted in 2008 and is holding up fairly well with a few areas\nwhere the top layer of paint is starting to peel.\nA'}

In [338]:
updated_components = []

for component in structured_components:
        # Create a copy of the component data
        updated_component = component.copy()
        
        # Extract the component name without the ID and material prefixes
        component_name = extract_component_name(component["text"])
        component_variations = generate_component_variations(component_name)
        
        # Try to find a match in the section headers first
        header_match = find_in_headers(component_variations, desc_sections)
        
        if header_match:
            # If found in a header, add the section information
            updated_component["description_section"] = header_match
        else:
            # If not found in headers, try to find mentions in the content
            content_match = find_in_content(component_variations, desc_sections)
            
            if content_match:
                updated_component["description_section"] = content_match
            else:
                updated_component["description_section"] = {}
        
        updated_components.append(updated_component)
updated_components

[{'text': '107 Steel Open Girder',
  'full_section': '107 Steel Open Girder/Beam LF 2820 2290 200 330 0\n1000 Corrosion LF 530 0 200 330 0\n515 Steel Protective Coating SF 25255 25255 0 0 0',
  'defects': [{'id': '1000',
    'description': 'Corrosion',
    'unit': 'LF',
    'total': 530,
    'cs_values': [0, 200, 330, 0]},
   {'id': '515',
    'description': 'Steel Protective Coating',
    'unit': 'SF',
    'total': 25255,
    'cs_values': [25255, 0, 0, 0]}],
  'id': '107',
  'units': 'LF',
  'values': [2820, 2290, 200, 330, 0],
  'description_section': {'header': '59 - Superstructure',
   'condition_text': '6 - SATISFACTORY CONDITION - structural elements show some minor deterioration.',
   'condition_rating': 'SATISFACTORY',
   'assessment': 'structural elements show some minor deterioration.',
   'content': 'Seven (7) painted steel rolled beams are in satisfactory condition having scattered minor to moderate previous pitting\nalong the flanges and webs throughout. Superstructure was

## Get maintenance reccomendations

In [339]:
doc4

Route VT15 /  

Structure #00020 /  (Routine)

VT 00015 ML over LAMOILLE RIVER

Team Lead: Stephen Piro,  Inspection Date: 09/27/2023

Town: 41 - CAMBRIDGE

District 8, 15 - LAMOILLE County

Owner: 1 - State Highway Agency

Maintenance Responsibility: 1 - State Highway Agency


Route VT15 /  

Structure #00020 /  (Routine)

VT 00015 ML over LAMOILLE RIVER

Team Lead: Stephen Piro,  Inspection Date: 09/27/2023

Location: 0.9 MI E JCT. VT.104 N

Copyright © 2024 Microsoft and its suppliers. All rights reserved.

44.64779, -72.87252


Route VT15 /  

Structure #00020 /  (Routine)

VT 00015 ML over LAMOILLE RIVER

Team Lead: Stephen Piro,  Inspection Date: 09/27/2023

IDENTIFICATION

CLASSIFICATION

(1) State Names
(8) Structure Number
(5) Inventory Route
(2) Highway Agency District
(3) County Code
(4) Place Code
(6) Features Intersected
(7) Facility Carried
(9) Location
(11) Mile Point
(12) Base Highway Network
(13) LRS Inventory Rte & Subrte
(16) Latitude
(17) Longitude
(98) Border Bridg

### Get maintenance section boundaries

In [340]:
def find_maintenance_sections(doc):
    """
    Find all "Maintenance Needs" sections in the document, from the header to "Remarks".
    
    Args:
        doc: spaCy Doc object
        
    Returns:
        list: List of section text strings
    """
    text = doc.text
    
    # Pattern to match "Maintenance Needs" header
    maintenance_pattern = r'Maintenance\s+Needs'
    
    # Pattern to match "Remarks" header
    remarks_pattern = r'Remarks'
    
    # Find all occurrences of "Maintenance Needs"
    maintenance_matches = list(re.finditer(maintenance_pattern, text, re.IGNORECASE))
    
    sections = []
    
    for i, match in enumerate(maintenance_matches):
        # Find the start of this maintenance section
        section_start = match.start()
        
        # Find the end of this maintenance section (next "Remarks" or next "Maintenance Needs" or end of text)
        # First look for "Remarks" after this match
        remarks_match = re.search(remarks_pattern, text[section_start:], re.IGNORECASE)
        if remarks_match:
            section_end = section_start + remarks_match.start()
        else:
            # If no "Remarks" found, check if there's another "Maintenance Needs"
            if i < len(maintenance_matches) - 1:
                section_end = maintenance_matches[i + 1].start()
            else:
                # If this is the last maintenance section, go to the end of the text
                section_end = len(text)
        
        # Extract the section text
        section_text = text[section_start:section_end].strip()
        sections.append(section_text)
    
    return sections

In [341]:
def find_maintenance_sections(doc):
    """
    Find all "Maintenance Needs" sections in the document, from the header to "Remarks".
    
    Args:
        doc: spaCy Doc object
        
    Returns:
        list: List of section text strings
    """
    text = doc.text
    
    # Pattern to match "Maintenance Needs" header
    maintenance_pattern = r'Maintenance\s+Needs'
    
    # Pattern to match "Remarks" header
    remarks_pattern = r'Remarks'
    
    # Find all occurrences of "Maintenance Needs"
    maintenance_matches = list(re.finditer(maintenance_pattern, text, re.IGNORECASE))
    
    sections = []
    
    for i, match in enumerate(maintenance_matches):
        # Find the start of this maintenance section
        section_start = match.start()
        
        # Find the end of this maintenance section (next "Remarks" or next "Maintenance Needs" or end of text)
        # First look for "Remarks" after this match
        remarks_match = re.search(remarks_pattern, text[section_start:], re.IGNORECASE)
        
        if remarks_match:
            section_end = section_start + remarks_match.start()
        else:
            # If no "Remarks" found, check if there's another "Maintenance Needs"
            if i < len(maintenance_matches) - 1:
                section_end = maintenance_matches[i + 1].start()
            else:
                # If this is the last maintenance section, go to the end of the text
                section_end = len(text)
        
        # Extract the section text
        section_text = text[section_start:section_end].strip()
        sections.append(section_text)
    
    return sections

In [343]:
maintenance_sections = find_maintenance_sections(doc4)
len(maintenance_sections)

2

In [344]:
for mainten in maintenance_sections:
    print(mainten)

Maintenance Needs

Date Reported: 08/23/2023

Priority:

Flood Event

Status:

Open

Type of Work:

38 - Channel - Scour/Undermining repair

Component:

Channel

Deficiency Description

Pier 3 footing along the upstream end has localized scouring that has exposed the bottom of footing.  Large stone should 
be added to protect this area so that scour does not continue grow in this area.
Maintenance Needs

Date Reported: 09/27/2023

Priority:

4 - Maintenance Finding - Next 
Inspection Cycle

Status:

Open

Type of Work:

46 - Deck - Joint cleaning

Component:

Deck

Deficiency Description

Both vermont type joint troughs over pier #1 and #2 are full of debris and sediment with vegetation growth and are in need 
of cleaning.


### Parse maintenance section boundaries

In [345]:
def extract_maintenance_info(maintenance_sections: List[str]) -> List[Dict]:
    """
    Extract maintenance information from an array of text sections
    
    Args:
        maintenance_sections: List of text strings, each representing a maintenance section
        
    Returns:
        List of dictionaries containing extracted maintenance information
    """
    # List to store all processed maintenance sections
    processed_sections = []
    
    # Process each section
    for section_text in maintenance_sections:
        current_section = {}
        current_section["full_text"] = section_text
        # Extract date reported
        date_match = re.search(r"Date Reported:\s*(\d{1,2}/\d{1,2}/\d{4})", section_text)
        if date_match:
            current_section["date_reported"] = date_match.group(1).strip()
        
        # Extract priority - look for content between "Priority:" and the next field
        priority_match = re.search(r"Priority:(?:\s*\n)?(.*?)(?=\n\s*(?:Type of Work:|Status:|Component:))", section_text, re.DOTALL)
        if priority_match:
            priority_text = priority_match.group(1).strip()
            # Clean up multi-line text
            priority_text = re.sub(r'\s+', ' ', priority_text)
            current_section["priority"] = priority_text
        
        # Extract type of work - look for content between "Type of Work:" and the next field
        work_match = re.search(r"Type of Work:(?:\s*\n)?(.*?)(?=\n\s*(?:Status:|Component:|Deficiency))", section_text, re.DOTALL)
        if work_match:
            work_text = work_match.group(1).strip()
            # Clean up multi-line text
            work_text = re.sub(r'\s+', ' ', work_text)
            current_section["type_of_work"] = work_text
        
        # Extract status
        status_match = re.search(r"Status:(?:\s*\n)?(.*?)(?=\n\s*(?:Component:|Deficiency))", section_text, re.DOTALL)
        if status_match:
            status_text = status_match.group(1).strip()
            current_section["status"] = status_text
        
        # Extract component
        component_match = re.search(r"Component:(?:\s*\n)?(.*?)(?=\n\s*(?:Deficiency|Remarks))", section_text, re.DOTALL)
        if component_match:
            component_text = component_match.group(1).strip()
            current_section["component"] = component_text
        
        # Try multiple patterns for deficiency description, just like in the first code
        deficiency_patterns = [
            r"Deficiency Description\s*(.+?)(?=Remarks:|$)",
            r"Deficiency Description\n(.+?)(?=\n\n|\n[A-Z]|$)"
        ]
        
        for pattern in deficiency_patterns:
            deficiency_match = re.search(pattern, section_text, re.DOTALL | re.IGNORECASE)
            if deficiency_match:
                deficiency_text = deficiency_match.group(1).strip()
                # Clean up multi-line text
                deficiency_text = re.sub(r'\s+', ' ', deficiency_text)
                current_section["deficiency_description"] = deficiency_text
                break  # Found a match, no need to try other patterns
        
        # Only add sections that have at least some data
        if current_section:
            processed_sections.append(current_section)
    
    return processed_sections

In [346]:
print(maintenance_sections[0])

Maintenance Needs

Date Reported: 08/23/2023

Priority:

Flood Event

Status:

Open

Type of Work:

38 - Channel - Scour/Undermining repair

Component:

Channel

Deficiency Description

Pier 3 footing along the upstream end has localized scouring that has exposed the bottom of footing.  Large stone should 
be added to protect this area so that scour does not continue grow in this area.


In [347]:
type(maintenance_sections[0])

str

In [348]:
maintenance_sections

['Maintenance Needs\n\nDate Reported: 08/23/2023\n\nPriority:\n\nFlood Event\n\nStatus:\n\nOpen\n\nType of Work:\n\n38 - Channel - Scour/Undermining repair\n\nComponent:\n\nChannel\n\nDeficiency Description\n\nPier 3 footing along the upstream end has localized scouring that has exposed the bottom of footing.  Large stone should \nbe added to protect this area so that scour does not continue grow in this area.',
 'Maintenance Needs\n\nDate Reported: 09/27/2023\n\nPriority:\n\n4 - Maintenance Finding - Next \nInspection Cycle\n\nStatus:\n\nOpen\n\nType of Work:\n\n46 - Deck - Joint cleaning\n\nComponent:\n\nDeck\n\nDeficiency Description\n\nBoth vermont type joint troughs over pier #1 and #2 are full of debris and sediment with vegetation growth and are in need \nof cleaning.']

In [349]:
mainntenance_needs = extract_maintenance_info(maintenance_sections)
mainntenance_needs

[{'full_text': 'Maintenance Needs\n\nDate Reported: 08/23/2023\n\nPriority:\n\nFlood Event\n\nStatus:\n\nOpen\n\nType of Work:\n\n38 - Channel - Scour/Undermining repair\n\nComponent:\n\nChannel\n\nDeficiency Description\n\nPier 3 footing along the upstream end has localized scouring that has exposed the bottom of footing.  Large stone should \nbe added to protect this area so that scour does not continue grow in this area.',
  'date_reported': '08/23/2023',
  'priority': 'Flood Event',
  'type_of_work': '38 - Channel - Scour/Undermining repair',
  'status': 'Open\n\nType of Work:\n\n38 - Channel - Scour/Undermining repair',
  'component': 'Channel',
  'deficiency_description': 'Pier 3 footing along the upstream end has localized scouring that has exposed the bottom of footing. Large stone should be added to protect this area so that scour does not continue grow in this area.'},
 {'full_text': 'Maintenance Needs\n\nDate Reported: 09/27/2023\n\nPriority:\n\n4 - Maintenance Finding - Nex

## Component and maintenance needs matching

In [350]:
updated_components_new = []
for component in updated_components:
        updated_component = component.copy()
        updated_component["maintenance_needs"] = []
        updated_components_new.append(updated_component)
updated_components_new

[{'text': '107 Steel Open Girder',
  'full_section': '107 Steel Open Girder/Beam LF 2820 2290 200 330 0\n1000 Corrosion LF 530 0 200 330 0\n515 Steel Protective Coating SF 25255 25255 0 0 0',
  'defects': [{'id': '1000',
    'description': 'Corrosion',
    'unit': 'LF',
    'total': 530,
    'cs_values': [0, 200, 330, 0]},
   {'id': '515',
    'description': 'Steel Protective Coating',
    'unit': 'SF',
    'total': 25255,
    'cs_values': [25255, 0, 0, 0]}],
  'id': '107',
  'units': 'LF',
  'values': [2820, 2290, 200, 330, 0],
  'description_section': {'header': '59 - Superstructure',
   'condition_text': '6 - SATISFACTORY CONDITION - structural elements show some minor deterioration.',
   'condition_rating': 'SATISFACTORY',
   'assessment': 'structural elements show some minor deterioration.',
   'content': 'Seven (7) painted steel rolled beams are in satisfactory condition having scattered minor to moderate previous pitting\nalong the flanges and webs throughout. Superstructure was

In [351]:
def match_by_component_name(need, components):
    """
    Match a maintenance need to a component by direct component name comparison.
    
    Args:
        need (dict): Maintenance need dictionary
        components (list): List of component dictionaries
        
    Returns:
        dict or None: Matched component or None if no match found
    """
    if not need.get("component"):
        return None
    
    need_component = need["component"].lower()
    
    for component in components:
        # Extract clean component name
        clean_name = extract_component_name(component["text"]).lower()
        
        # Check for direct match or substring match
        if (clean_name == need_component or 
            clean_name in need_component or 
            need_component in clean_name):
            return component
        
        # Handle special cases
        if clean_name == "wing/retaining wall" and ("wing" in need_component or "retaining" in need_component or "wall" in need_component):
            return component
        
        if clean_name == "abutment" and "abutment" in need_component:
            return component
    
    return None

In [143]:
def match_by_description(need, components):
    """
    Match a maintenance need to a component by finding component mentions in the description.
    
    Args:
        need (dict): Maintenance need dictionary
        components (list): List of component dictionaries
        
    Returns:
        dict or None: Matched component or None if no match found
    """
    if not need.get("deficiency_description"):
        return None
    
    description = need["deficiency_description"].lower()
    
    # Try each component
    for component in components:
        # Extract clean component name and variations
        clean_name = extract_component_name(component["text"])
        variations = generate_component_variations(clean_name)
        
        # Check if any variation appears in the description
        for variation in variations:
            pattern = r'\b' + variation.lower() + r'\b'
            if re.search(pattern, description):
                return component
            
            # Handle numbered components like "abutment 1"
            if variation.lower() + r' \d+' in variations:
                pattern = r'\b' + variation.lower() + r' \d+\b'
                if re.search(pattern, description):
                    return component
    
    return None


In [352]:
# Process each maintenance need
for need in mainntenance_needs:
        # Try to match by direct component name
    matched_component = match_by_component_name(need, updated_components_new)
        
        # If no match, try to match by mentions in the deficiency description
    if not matched_component:
        matched_component = match_by_description(need, updated_components_new)
        
        # If a match was found, add the maintenance need to the component
    if matched_component:
        matched_component["maintenance_needs"].append(need)
    
updated_components_new

[{'text': '107 Steel Open Girder',
  'full_section': '107 Steel Open Girder/Beam LF 2820 2290 200 330 0\n1000 Corrosion LF 530 0 200 330 0\n515 Steel Protective Coating SF 25255 25255 0 0 0',
  'defects': [{'id': '1000',
    'description': 'Corrosion',
    'unit': 'LF',
    'total': 530,
    'cs_values': [0, 200, 330, 0]},
   {'id': '515',
    'description': 'Steel Protective Coating',
    'unit': 'SF',
    'total': 25255,
    'cs_values': [25255, 0, 0, 0]}],
  'id': '107',
  'units': 'LF',
  'values': [2820, 2290, 200, 330, 0],
  'description_section': {'header': '59 - Superstructure',
   'condition_text': '6 - SATISFACTORY CONDITION - structural elements show some minor deterioration.',
   'condition_rating': 'SATISFACTORY',
   'assessment': 'structural elements show some minor deterioration.',
   'content': 'Seven (7) painted steel rolled beams are in satisfactory condition having scattered minor to moderate previous pitting\nalong the flanges and webs throughout. Superstructure was

In [145]:
len(updated_components_new)

3